src/mem/secure.c

```diff
@@ -35,3 +45,35 @@ int mem_seccmp(const uint8_t *s1, const uint8_t *s2, size_t n)
 
 	return val;
 }
+
+
+#if !defined(__GNUC__) && !defined(USE_OPENSSL) && !defined(WIN32)
+/* Use a volatile pointer to memset to force the compiler always
+ * call it and not optimize away. */
+typedef void *(memset_t)(void *, int, size_t);
+static memset_t *const volatile memset_ptr = &memset;
+#endif
+
+/**
+ * Securely clean memory. This function is guaranteed not to get optimized
+ * away by compiler.
+ *
+ * @param data Pointer to data buffer
+ * @param size Size of the buffer
+ */
+void mem_secclean(void *data, size_t size)
+{
+#if defined(__GNUC__)
+	memset(data, 0, size);
+	/* Insert an asm statement that may potentially depend
+	* on the memory contents that were affected by memset.
+	* This prevents optimizing away the memset. */
+	__asm__ __volatile__("" : : "r" (data), "r" (size) : "memory");
```

Comment:
Does this offer more guarantees than the `(*memset_ptr)(data, 0, size);` variant?

Reply:
No, but it eliminates a volatile load and likely improves branch prediction.

Reply:
Thanks for clarification, nice.